### ISSUE TITLE
GPIO: If a driver implements both blocking and async operations, or only implements blocking operations, but may support asynchronous ones in the future, the driver's type signature must include a `crate::Mode` type parameter.

### ISSUE THREAD

**Issue by bugadani** [esp-rs/esp-hal#2709]
- By default, constructors must configure the driver for blocking mode. The driver must implement into_async (and a matching into_blocking) function that reconfigures the driver.
- into_async must configure the driver and/or the associated DMA channels. This most often means enabling an interrupt handler.
- into_blocking must undo the configuration done by into_async.

Note that `Avoid type states and extraneous generics whenever possible` gives us a possible justification to just close this issue without changes.

**Comment by MabezDev** [comment]
I don't think we want to add `Mode` parameters to pins, I think this is an (the only?) exception to this guideline.

Closing.
